File: Virtual Machine/Nova.VirtualMachine/Nova.VirtualMachine/ByteClass.cpp
#include "ByteClass.h"

void ByteClass::Deserialize(BinaryReader& reader)
{
	this->name = reader.ReadString();

	this->type = (ContainerType)reader.Read<char>();

	int methodsCount = reader.Read<int>();

	for (int i = 0; i < methodsCount; i++)
	{
		ByteMethod* method = new ByteMethod(this);
		method->Deserialize(reader);
		this->methods.push_back(method);
	}

	int fieldsCount = reader.Read<int>();

	for (int i = 0; i < fieldsCount; i++)
	{
		ByteField* field = new ByteField();
		field->Deserialize(reader);
		this->fields.push_back(field);
	}

	
}

void ByteClass::Dispose()
{
	for (ByteMethod* method : this->methods)
	{
		method->Dispose();
		delete method;
	}
	for (ByteField* field : this->fields)
	{
		field->Dispose();
		delete field;
	}
	
}

File: Compiler/Nova.VirtualMachine/Members/ByteClass.cs
using Nova.Utils.IO;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Nova.VirtualMachine.Members
{
    public class ByteClass
    {
        private string Name
        {
            get;
            set;
        }
        public List<ByteMethod> Methods
        {
            get;
            private set;
        }
        public List<ByteField> Fields
        {
            get;
            private set;
        }
        public List<object> Constants
        {
            get;
            private set;
        }
        public ByteClass()
        {
            this.Methods = new List<ByteMethod>();
            this.Fields = new List<ByteField>();
            this.Constants = new List<object>();
        }
        public void Deserialize(CppBinaryReader reader)
        {
            this.Name = reader.ReadString();

            int methodsCount = reader.ReadInt32();

            for (int i = 0; i < methodsCount; i++)
            {
                ByteMethod method = new ByteMethod(this);
                method.Deserialize(reader);
                Methods.Add(method);
            }

            int fieldsCount = reader.ReadInt32();

            for (int i = 0; i < fieldsCount; i++)
            {
                ByteField field = new ByteField();
                field.Deserialize(reader);
                Fields.Add(field);
            }

            int constantsCount = reader.ReadInt32();
            for (int i = 0; i < constantsCount; i++)
            {
                int type = reader.ReadInt32();

                switch (type)
                {
                    case 1:
                        Constants.Add(reader.ReadString());
                        break;
                    case 2:
                        Constants.Add(reader.ReadBoolean());
                        break;
                    default:
                        throw new Exception("Unknown constant type.");
                }
            }
        }
    }
}
